### ISSUE TITLE
CFSSL with NetHSM

### ISSUE THREAD

**Issue by tendaworld** [Nitrokey/nethsm-pkcs11#151]
With this mod applied, one can use PKCS https://github.com/cloudflare/cfssl/pull/11 token in multirootca for certs signing (tested succesfully with Nitrokey HSM 2; should
work with other compatible HSM tokens too). Requires importing

https://github.com/letsencrypt/pkcs11key

Example config below works with Nitrokey-HSM

[ default ]
private = pkcs11://notused
module = /usr/lib/x86_64-linux-gnu/opensc-pkcs11.so
token_label = UserPIN (myhsm)
pin = 123456
config = /etc/cfssl/server.json
certificate = /etc/cfssl/issuer.crt
Note: issuer.crt must contain cert issued for signing key in HSM.

However when using it on NetHSM like below, running into [FATAL] pkcs11key: pkcs11key: opening session: pkcs11: 0x6: CKR_FUNCTION_FAILED. Can you replicate the issue and assist with debugging.

[ default ]
private = pkcs11://notused
module = /usr/lib/nitrokey/libnethsm_pkcs11.so
token_label = pkcs11:object=MYKEY1;type=public
pin = ""
config = /etc/cfssl/server.json
certificate = /etc/cfssl/issuer.crt

**Comment by jans23** [comment]
Is your NetHSM provisioned and operational?

**Comment by tendaworld** [comment]
Yes, both from Rest all's good status operational, and also used openssl engine with pkcs11 to generate the root cert and imported it into the HSM.

**Comment by q-nk** [comment]
I do not fully understand your setup, so let me describe it in my own words:
You are using CFSSL to build certificate bundles. The keypair for the root certificate, of the bundles, should be stored within a HSM, accessed through PKCS#11, which was made available for CFSSL through this patch https://github.com/tenda-dev/cfssl?

I assume the mentioned code 
```
	// initialize token communication
	priv, err = pkcs11key.New(module, tokenLabel, pin, cert.PublicKey)
	if err != nil {
		return nil, err
	}
```
is from https://github.com/tenda-dev/cfssl/blob/4513b7df7b567b487ba1fd1f59d48b91b1798d7a/multiroot/config/config.go#L292?
This code parses the configs you posted? e.g.
```
[ default ]
private = pkcs11://notused
module = /usr/lib/nitrokey/libnethsm_pkcs11.so
token_label = pkcs11:object=MYKEY1;type=public
pin = ""
config = /etc/cfssl/server.json
certificate = /etc/cfssl/issuer.crt
```

Did you made any changes between your two posts? The error messages differ:
First Post:
```
[FATAL] pkcs11key: pkcs11key: opening session: pkcs11: 0x6: CKR_FUNCTION_FAILED
```
Second Post:
```
[FATAL] pkcs11key: pkcs11key: opening session: pkcs11: 0x7: CKR_ARGUMENTS_BAD
```

Could you provide a step-by-step manual to reproduce the mentioned behavior? 

My first impression is that some PKCS#11 parameters are wrong. So in the best case only the mentioned config has to be adjusted.

**Comment by tendaworld** [comment]
I was able to debug between the first and second post, found there was a wrong nethsm url, which I corrected.
The step to reproduce is using the patched CFSSL multiroot from here https://github.com/tenda-dev/cfssl/blob/4513b7df7b567b487ba1fd1f59d48b91b1798d7a/multiroot/config/config.go#L292
and using the roots.conf as below
[ default ]
private = pkcs11://notused
module = /usr/lib/nitrokey/libnethsm_pkcs11.so
token_label = LocalHSM
pin = ""
config = /etc/cfssl/ca_root.config.json
certificate = /etc/cfssl/CA-ROOT.crt

started CFSSL with the multirootca cmd
multirootca \
            -a 0000:8080 \
            -l default \
            -roots roots.conf \
            -tls-cert server-tls.pem \
            -tls-key server-tls-key.pem

I end up with the error 
[FATAL] pkcs11key: pkcs11key: opening session: pkcs11: 0x7: CKR_ARGUMENTS_BAD

I have created the pub/priv (CA-ROOT.pem) key pair on the nethsm using the REST API. 
I used openssl to generate a CSR and it's PKCS#11 engine to connect to the same nethsm and successfully create the self-signed CA-ROOT.crt. which I imported into the nethsm using the REST API.

I suspect the issue lies in how the CFSSL PKCS#11 driver is calling/sending instructions to the nethsm and without a debugging/pkcs#11spy or something to see how it is currently implemented i'm chasing a black cat in the dark. Perhaps you can help.to resolve.
